### DISCUSSION TITLE
what data type can transforms apply to? cv2 image? numpy array? other?

### DISCUSSION THREAD

**Discussion by xiaoyuan0203** [albumentations-team/albumentations#1217] [Q&A]
what data type can transforms apply to? cv2 image? numpy array? other?

**Comment by Dipet** [comment]
Numpy array in HWC format

**Reply by xiaoyuan0203** [reply]
Gray format? Or multispectral images?

**Reply by Dipet** [reply]
The gray image might be in HW format or HWC where C=1.
Almost all transforms can work on images with any number of channels.

**Reply by xiaoyuan0203** [reply]
Can it apply to numpy array with any channel?

**Reply by Dipet** [reply]
Yes, many transforms can work with any channel.  Exception - some specific transforms, that works only with RGB images, for example RGBShift

**Reply by xiaoyuan0203** [reply]
Dose a certain transform will indicate what type of image(grey, RGB, multichannel) it can apply to? and where?

**Reply by Dipet** [reply]
It looks like we really have a problem with this and we need to add this information to the docs.
Now, looks like the only way to be certain that transforms work with any count of channels - just check single transform.
All [spatial-level](https://albumentations.ai/docs/api_reference/full_reference/#spatial-level-transforms) transforms must work with any number of channels.
Main problems with transforms, that works on color: color shifts, hue, saturation, weather augs, etc.

**Reply by xiaoyuan0203** [reply]
I have an idea. We can add tags for each transform indicating the type of this transform. The type of transform can be:
- pixel-level: not change the position information of the image
- spatial-level: change the position information of the image
- channel-independent: apply to each channel independently(so it can apply to any number of channels of cource), such as translate, rotate...
- channel-dependent: such as color shift

**Reply by xiaoyuan0203** [reply]
And make a function to filter certein types of transforms as a search space so that kinds of composed augmentation methods such as augmix can make a transform-chain from certain types of tranforms. Augmix and other composed augmentation methods can integrated to this repo with the existing transforms

**Reply by xiaoyuan0203** [reply]
These methods don't make specific transform but offer a composition policy to compose the existing transforms. The simplest composition policy is: Compose all the transforms offered sequentially. Augmix is doing such a task. Other policy can be openly explored.
